### PR TITLE
Inverse quadratic approximation for MappingQ::real_to_unit_cell

### PR DESCRIPTION
**Pull request by kronbichler**
~~Only the last commit is relevant, the rest is #11054.~~

This PR introduces a quadratic approximation of the inverse map of a MappingQGeneric (which is generally a rational function) and uses it for `MappingQGeneric::transform_points_real_to_unit_cell`. I only use it for the function computing multiple points because the least squares fit takes some effort to compute, something on the order of three Newton iterations of the `do_transform_real_to_unit_cell_internal`. The approximation reduces the number of Newton iterations by about 0.7 iterations seen as the average over many tests, so it seems worth it.

### CODE REVIEW

File: tests/mappings/mapping_q_inverse_quadratic_approximation.output [hunk 1]
```diff
@@ -0,0 +1,355 @@
+
+DEAL::Testing 2D with point -0.1498310826 0.7854420410
+DEAL::Testing on cell 0_1:0 with center 0.8227241336 0.4750000000
+DEAL::Exact inverse:                   1.679734433 2.008122269
+DEAL::Affine approximation:            1.294871897 6.963559088
+DEAL::Inverse quadratic approximation: 1.644896745 2.003800318
```
Comment (drwells) [inline tests/mappings/mapping_q_inverse_quadratic_approximation.output:6]: That's a pretty good approximation!

**Review by drwells — CHANGES_REQUESTED** [review]
These are some great results - I only have some minor documentation questions.

File: include/deal.II/fe/mapping_q_internal.h [hunk 2]
```diff
@@ -811,6 +813,202 @@ namespace internal
...
+    {
+    public:
+      /**
+       * Number of basis functions in the quadratic approximation.
+       */
+      static constexpr unsigned int n_functions =
+        (spacedim == 1 ? 3 : (spacedim == 2 ? 6 : 10));
+
+      /**
+       * Constructor.
```
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/deal.II/fe/mapping_q_internal.h:846]: These two arguments are used to construct the reference points corresponding to the support points in real space. One might argue that it is more appropriate to pass in the reference points directly to make the class more independent of `MappingQGeneric` and make the linear algebra point of view clear - I match the function from the real points to a right hand side in the reference points. (I could see this being used for the inverse finding of `TransfiniteInterpolationManifold` as well.)

File: tests/mappings/mapping_q_inverse_quadratic_approximation.output (hunk 1 above)
Comment (kronbichler) [reply tests/mappings/mapping_q_inverse_quadratic_approximation.output:6]: A quarter of a annulus is also something that is approximated badly with an affine approximation whereas a quadratic function captures almost all of it.

File: include/deal.II/fe/mapping_q_internal.h [hunk 3]
```diff
@@ -811,6 +813,202 @@ namespace internal
...
+       */
+      template <typename Number>
+      Point<dim, Number>
+      compute(const Point<spacedim, Number> &p)
+      {
+        Point<dim, Number> result;
+        for (unsigned int d = 0; d < dim; ++d)
+          result[d] = coefficients[0][d];
+
+        // Shift point to avoid roundoff problems. Spell out the loop
```
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/deal.II/fe/mapping_q_internal.h:984]: Let me rephrase the 'spell' part - I simply want to say that I cannot use the overloads of `Point` but must write the loop over the `spacedim` components.

File: include/deal.II/fe/mapping_q_internal.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/deal.II/fe/mapping_q_internal.h:874]: I took the chance to switch the constructor to the intuitive setup: I specify the points in real space and the matching points in unit space, which are used to make the least squares fitting.

**Comment by kronbichler** [comment]
I addressed all comments.

**Review by drwells — APPROVED** [review]
